### ISSUE TITLE
feature request save states

### ISSUE THREAD

**Issue by InfamousKnight** [ruffle-rs/ruffle#1446]

**Comment by Dinnerbone** [comment]
Can you please elaborate a little further on this? What is the functionality you want and how would you imagine it working?

**Comment by InfamousKnight** [comment]
Save states are snapshots of the game. It saves everything in a file. The
user can then load the save state.

And by everything, I mean every last register, ram, timer you name it. It's
a common emulator feature and would be a big undertaking. I dont know how
much work would be needed with flash though

**Comment by thecnoNSMB** [comment]
I would like to note for the record that this feature is a must for tool-assisted speedrunners, who are a notable fraction of emulator users.

**Comment by Lord-McSweeney** [comment]
#10040 added a save manager.

**Comment by Dinnerbone** [comment]
That's not what's being asked for here.

**Comment by torokati44** [comment]
I agree, so reopening.

**Comment by EnergyBun** [comment]
Forwarded from #18667 :

Ruffle is emulation for Flash, and all other forms of emulation have Saving and Loading States, so Flash should too. The benefits of Save States for Flash are the exact same for Saves States for other game emulation, but I will list them here to be clear:

First and foremost, Saves States will give the longest and most deserving Flash games the chance to be enjoyed by people who cannot play the entire game in one sitting. If you need evidence for this, you need look no further than Portal the Flash Version. Videos play-throughs of this game are 20-48 minutes which is not necessarily counting failed attempts. It is possible to take breaks in this particular game by leaving it running, but that is not necessarily true for all games, for example those with timers. This also does not account for unexpectedly long absences or sudden power loss.

The second reason for a need of Saves States is physical and mental fatigue. Many Flash games of the early 2000s were designed to require extremely twitchy reflexes and precision. I cannot say whether this was caused by accident or simply the norm at the time, but suffice it to say not everyone is able to devote the time and energy to practice a reflex response for what may be say level seven of a game for days only for the game to be reset after an in-game defeat requiring the player to play all previous levels before having another attempt at the one that demands so much.

Thirdly, as a continuation of the above, users who simply find themselves unable to pass a certain point in a Flash game may ask their friends or other enthusiasts to pass a point for them by loading the Save State and sending one in return after the difficult point. Back in the day, some Flash games were played by groups of friends meeting at a single house and passing control as needed, but with Flash enthusiasts being spread out, meeting in person is hardly an option.

In conclusion, I see Flash emulation as no different than retro game emulation, deserving of all the decoration, celebration, and after-the-fact upgrades. The one thing that separates the experience of these two communities is that Flash games are currently in a state of play it all in one sitting or never reach the end. I hope that capable programmers will fix that.
